gitrelease/gitrelease.go

```diff
@@ -0,0 +1,446 @@
+	commitSHA256 := hex.EncodeToString(h)
+
+	remoteMain := plumbing.NewRemoteReferenceName("origin", "main")
+	remoteRef, err := repo.Reference(remoteMain, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed ot find origin/main: %w", err)
+	}
+
+	remoteMainHash := remoteRef.Hash().String()
+	h, err = gh.CommitSum(commitHash)
+	if err != nil {
+		return nil, err
+	}
+	remoteMainSHA256 := hex.EncodeToString(h)
+
+	goHashes, err := gohash.GitDirHashAllVersion(repo, commitHash, &tagName)
+	if err != nil {
+		return nil, err
+	}
+
+	var goRelease *GoRelease = nil
+	for _, goHash := range goHashes {
+		if goHash.Directory == "" {
+			goRelease = &GoRelease{
+				// https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#golang
+				Uri:           "pkg:golang/" + goHash.Path + "@" + goHash.Version,
+				DirectoryHash: goHash.DirectoryHash,
+				GoModHash:     goHash.GoModHash,
+			}
+		}
+	}
+
+	tagMetadata := TagMetadata{
+		Type:          "https://supply-chain-tools.github.io/schemas/tag-metadata/v0.0",
+		RepositoryUri: wrapGitURL(repoUrl),
+		CommitHash: CommitHash{
+			SHA1:   &commitSHA1,
+			SHA256: &commitSHA256,
+		},
+		PreviousTag: nil,
+		ProtectedBranches: []ProtectedBranch{
+			{
+				Name:   "main",
+				SHA1:   &remoteMainHash,
+				SHA256: &remoteMainSHA256,
+			},
+		},
+		GoRelease: goRelease,
+	}
+
+	if previousTag != nil {
+		gh := githash.NewGitHash(repo, sha256.New())
+		tagSHA256, err := gh.TagSum(previousTag.Hash)
+		if err != nil {
+			return nil, err
+		}
+
+		previousTagSHA1 := previousTag.Hash.String()
+		previousTagSHA256 := hex.EncodeToString(tagSHA256)
+		tagMetadata.PreviousTag = &PreviousTag{
+			SHA1:   &previousTagSHA1,
+			SHA256: &previousTagSHA256,
+		}
+	}
+
+	return &tagMetadata, nil
+}
+
+func FindPreviousTag(repo *git.Repository) (*object.Tag, bool, error) {
+	tags, err := repo.Tags()
+	if err != nil {
+		return nil, false, err
+	}
+
+	state := gitkit.LoadRepoState(repo)
+	var previousTag *object.Tag
+
+	tagHashes := make(map[string]*object.Tag)
+	previousTagHashes := make(map[string]*TagMetadata)
+
+	err = tags.ForEach(func(tag *plumbing.Reference) error {
+		t, isAnnotatedTag := state.TagMap[tag.Hash()]
+
+		if isAnnotatedTag {
+			tagMetadata, err := decodeTagMetadata(t)
+			if err != nil {
+				return err
+			}
+
+			tagHashes[tag.Hash().String()] = t
+
+			if tagMetadata.PreviousTag != nil {
+				// TODO also verify SHA-256
+				targetHash := tagMetadata.PreviousTag.SHA1
+				_, found := previousTagHashes[*targetHash]
+				if found {
+					return fmt.Errorf("duplicate previous tag '%s'", *targetHash)
+				}
+
+				previousTagHashes[*targetHash] = tagMetadata
+			}
+		} else {
+			return fmt.Errorf("lightweight tags are not supported %s", tag.Hash().String())
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, false, err
+	}
+
+	count := 0
+	for hash, tag := range tagHashes {
+		if _, found := previousTagHashes[hash]; !found {
+			previousTag = tag
+			count++
+		}
+	}
```

Comment:
Was a bit stuck on this one for a moment, maybe add a comment?

```suggestion
	count := 0
	for hash, tag := range tagHashes {
	        // When we find a tag not pointed to by others we have a latest tag
	        // candidate. There may be more tags with metadata, which should
	        // not happen, so we keep track with a count.
		if _, found := previousTagHashes[hash]; !found {
			previousTag = tag
			count++
		}
	}
```